File: app/src/main/java/ru/vasiliev/sandbox/camera2/device/camera/Camera2RequestManager.java
package ru.vasiliev.sandbox.camera2.device.camera;

import android.hardware.camera2.CameraAccessException;
import android.hardware.camera2.CameraDevice;
import android.hardware.camera2.CaptureRequest;
import android.view.Surface;

import androidx.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;

import ru.vasiliev.sandbox.camera2.device.camera.util.CameraFacing;
import ru.vasiliev.sandbox.camera2.device.camera.util.CameraFlash;

public class Camera2RequestManager {

    private CameraDevice cameraDevice;

    private Camera2Config camera2Config;

    Camera2RequestManager(CameraDevice cameraDevice, Camera2Config camera2Config) {
        this.cameraDevice = cameraDevice;
        this.camera2Config = camera2Config;
    }

    private void setup3AControls(CaptureRequest.Builder requestBuilder, boolean autoFocus, CameraFlash cameraFlash,
                                 boolean autoWhiteBalance) {
        // Enable auto-magical 3A run
        requestBuilder.set(CaptureRequest.CONTROL_MODE, CaptureRequest.CONTROL_MODE_AUTO);

        // Auto focus
        if (autoFocus && camera2Config.isAfSupported()) {
            requestBuilder.set(CaptureRequest.CONTROL_AF_MODE, camera2Config.getOptimalAfMode());
        } else {
            requestBuilder.set(CaptureRequest.CONTROL_AF_MODE, CaptureRequest.CONTROL_AF_MODE_OFF);
        }

        // Flash and auto exposure
        requestBuilder.set(CaptureRequest.CONTROL_AE_MODE, cameraFlash.getAeMode());
        requestBuilder.set(CaptureRequest.FLASH_MODE, cameraFlash.getFlashMode());

        // Auto white balance
        if (autoWhiteBalance && camera2Config.isAwbSupported()) {
            requestBuilder.set(CaptureRequest.CONTROL_AWB_MODE, CaptureRequest.CONTROL_AWB_MODE_AUTO);
        } else {
            requestBuilder.set(CaptureRequest.CONTROL_AWB_MODE, CaptureRequest.CONTROL_AWB_MODE_OFF);
        }
    }

    private void setupJpegOrientation(CaptureRequest.Builder requestBuilder, CameraFacing cameraFacing,
                                      int displayOrientation) {
        requestBuilder.set(CaptureRequest.JPEG_ORIENTATION,
                           (camera2Config.getCameraSensorOrientation() +
                            displayOrientation * (cameraFacing == CameraFacing.FRONT ? 1 : -1) + 360) % 360);
    }

    PreviewRequestBuilder newPreviewRequestBuilder() throws CameraAccessException {
        return new PreviewRequestBuilder();
    }

    CaptureRequestBuilder newCaptureRequestBuilder() throws CameraAccessException {
        return new CaptureRequestBuilder();
    }

    public class PreviewRequestBuilder {

        private CaptureRequest.Builder previewRequestBuilder;
        private List<Surface> outputSurfaces = new ArrayList<>();
        private boolean autoFocus = false;
        private CameraFlash cameraFlash = CameraFlash.FLASH_OFF;
        private boolean autoWhiteBalance = false;

        private PreviewRequestBuilder() throws CameraAccessException {
            previewRequestBuilder = cameraDevice.createCaptureRequest(CameraDevice.TEMPLATE_PREVIEW);
        }

        PreviewRequestBuilder setOutputSurface(Surface outputSurface) {
            outputSurfaces.add(outputSurface);
            return this;
        }

        PreviewRequestBuilder setOutputSurfaces(List<Surface> outputSurfaces) {
            this.outputSurfaces = outputSurfaces;
            return this;
        }

        PreviewRequestBuilder setAutoFocus(boolean autoFocus) {
            this.autoFocus = autoFocus;
            return this;
        }

        PreviewRequestBuilder setCameraFlash(CameraFlash cameraFlash) {
            this.cameraFlash = cameraFlash;
            return this;
        }

        PreviewRequestBuilder setAutoWhiteBalance(boolean autoWhiteBalance) {
            this.autoWhiteBalance = autoWhiteBalance;
            return this;
        }

        <T> PreviewRequestBuilder setKeyValue(@NonNull CaptureRequest.Key<T> key, T value) {
            previewRequestBuilder.set(key, value);
            return this;
        }

        public CaptureRequest build() {
            setup3AControls(previewRequestBuilder, autoFocus, cameraFlash, autoWhiteBalance);
            for (Surface surface : outputSurfaces) {
                previewRequestBuilder.addTarget(surface);
            }
            return previewRequestBuilder.build();
        }
    }

    public class CaptureRequestBuilder {

        private CaptureRequest.Builder previewRequestBuilder;
        private List<Surface> outputSurfaces = new ArrayList<>();
        private boolean autoFocus = false;
        private CameraFlash cameraFlash = CameraFlash.FLASH_OFF;
        private boolean autoWhiteBalance = false;
        private CameraFacing cameraFacing = CameraFacing.BACK;
        private int displayOrientation = Surface.ROTATION_0;

        private CaptureRequestBuilder() throws CameraAccessException {
            previewRequestBuilder = cameraDevice.createCaptureRequest(CameraDevice.TEMPLATE_STILL_CAPTURE);
        }

        CaptureRequestBuilder setOutputSurface(Surface outputSurface) {
            outputSurfaces.add(outputSurface);
            return this;
        }

        CaptureRequestBuilder setOutputSurfaces(List<Surface> outputSurfaces) {
            this.outputSurfaces = outputSurfaces;
            return this;
        }

        CaptureRequestBuilder setAutoFocus(boolean autoFocus) {
            this.autoFocus = autoFocus;
            return this;
        }

        CaptureRequestBuilder setCameraFlash(CameraFlash cameraFlash) {
            this.cameraFlash = cameraFlash;
            return this;
        }

        CaptureRequestBuilder setAutoWhiteBalance(boolean autoWhiteBalance) {
            this.autoWhiteBalance = autoWhiteBalance;
            return this;
        }

        CaptureRequestBuilder setDisplayOrientation(CameraFacing cameraFacing, int displayOrientation) {
            this.cameraFacing = cameraFacing;
            this.displayOrientation = displayOrientation;
            return this;
        }

        <T> CaptureRequestBuilder setKeyValue(@NonNull CaptureRequest.Key<T> key, T value) {
            previewRequestBuilder.set(key, value);
            return this;
        }

        public CaptureRequest build() {
            setup3AControls(previewRequestBuilder, autoFocus, cameraFlash, autoWhiteBalance);
            setDisplayOrientation(cameraFacing, displayOrientation);
            for (Surface surface : outputSurfaces) {
                previewRequestBuilder.addTarget(surface);
            }
            return previewRequestBuilder.build();
        }
    }
}

File: app/src/main/java/ru/vasiliev/sandbox/camera/device/camera/camera2/Camera2RequestManager.kt
package ru.vasiliev.sandbox.camera.device.camera.camera2

import android.hardware.camera2.CameraAccessException
import android.hardware.camera2.CameraDevice
import android.hardware.camera2.CaptureRequest
import android.view.Surface
import ru.vasiliev.sandbox.camera.device.camera.common.CameraPreview
import ru.vasiliev.sandbox.camera.device.camera.util.CameraFacing
import ru.vasiliev.sandbox.camera.device.camera.util.CameraFlash
import java.util.*

internal class Camera2RequestManager(
    private val camera2Controller: Camera2Controller,
    private val camera2Config: Camera2Config,
    private val cameraPreview: CameraPreview
) {

    private fun setup3AControls(
        requestBuilder: CaptureRequest.Builder, autoFocus: Boolean, cameraFlash: CameraFlash,
        autoWhiteBalance: Boolean
    ) {
        // Enable auto-magical 3A run
        requestBuilder.set(CaptureRequest.CONTROL_MODE, CaptureRequest.CONTROL_MODE_AUTO)

        // Auto focus
        if (autoFocus && camera2Config.isAfSupported) {
            requestBuilder.set(CaptureRequest.CONTROL_AF_MODE, camera2Config.optimalAfMode)
        } else {
            requestBuilder.set(CaptureRequest.CONTROL_AF_MODE, CaptureRequest.CONTROL_AF_MODE_OFF)
        }

        // Flash and auto exposure
        requestBuilder.set(CaptureRequest.CONTROL_AE_MODE, cameraFlash.aeMode)
        requestBuilder.set(CaptureRequest.FLASH_MODE, cameraFlash.flashMode)

        // Auto white balance
        if (autoWhiteBalance && camera2Config.isAwbSupported) {
            requestBuilder.set(
                CaptureRequest.CONTROL_AWB_MODE,
                CaptureRequest.CONTROL_AWB_MODE_AUTO
            )
        } else {
            requestBuilder.set(CaptureRequest.CONTROL_AWB_MODE, CaptureRequest.CONTROL_AWB_MODE_OFF)
        }
    }

    private fun setupJpegOrientation(
        requestBuilder: CaptureRequest.Builder, cameraFacing: CameraFacing,
        displayOrientation: Int
    ) {
        requestBuilder.set(
            CaptureRequest.JPEG_ORIENTATION,
            (camera2Config.cameraSensorOrientation + displayOrientation * if (cameraFacing == CameraFacing.FRONT) 1 else -1 + 360) % 360
        )
    }

    /**
     * @return Preview request builder with actual preview params from controller
     * - Auto-focus
     * - Flash
     * - Auto white-balance
     * @throws CameraAccessException when camera is not accessible
     */
    @Throws(CameraAccessException::class)
    fun newPreviewRequestBuilder(): PreviewRequestBuilder {
        return PreviewRequestBuilder()
    }

    /**
     * @return Capture request builder with actual preview params from linked controller:
     * - Auto-focus
     * - Flash
     * - Auto white-balance
     * - Camera facing
     * - Display rotation
     * @throws CameraAccessException when camera is not accessible
     */
    @Throws(CameraAccessException::class)
    fun newCaptureRequestBuilder(captureConfigProvider: CaptureConfigProvider): CaptureRequestBuilder {
        return CaptureRequestBuilder()
    }

    interface CaptureConfigProvider {
        var autoFocus: Boolean
        var cameraFlash: CameraFlash
        var autoWhiteBalance: Boolean
    }

    inner class PreviewRequestBuilder(private val captureConfigProvider: CaptureConfigProvider) {
        private val previewRequestBuilder: CaptureRequest.Builder =
            camera2Controller.cameraDevice.createCaptureRequest(CameraDevice.TEMPLATE_PREVIEW)
        private var outputSurfaces: MutableList<Surface> = ArrayList()
        private var autoFocus: Boolean = camera2Controller.getAutoFocus()
        private var cameraFlash: CameraFlash = camera2Controller.getFlash()
        private var autoWhiteBalance: Boolean = camera2Controller.getAutoWhiteBalance()
        fun setOutputSurface(outputSurface: Surface): PreviewRequestBuilder {
            outputSurfaces.add(outputSurface)
            return this
        }

        fun setOutputSurfaces(outputSurfaces: MutableList<Surface>): PreviewRequestBuilder {
            this.outputSurfaces = outputSurfaces
            return this
        }

        fun setAutoFocus(autoFocus: Boolean): PreviewRequestBuilder {
            this.autoFocus = autoFocus
            return this
        }

        fun setCameraFlash(cameraFlash: CameraFlash): PreviewRequestBuilder {
            this.cameraFlash = cameraFlash
            return this
        }

        fun setAutoWhiteBalance(autoWhiteBalance: Boolean): PreviewRequestBuilder {
            this.autoWhiteBalance = autoWhiteBalance
            return this
        }

        fun <T> setKeyValue(key: CaptureRequest.Key<T>, value: T): PreviewRequestBuilder {
            previewRequestBuilder[key] = value
            return this
        }

        fun build(): CaptureRequest {
            setup3AControls(previewRequestBuilder, autoFocus, cameraFlash, autoWhiteBalance)
            for (surface in outputSurfaces) {
                previewRequestBuilder.addTarget(surface)
            }
            return previewRequestBuilder.build()
        }

    }

    inner class CaptureRequestBuilder private constructor() {
        private val previewRequestBuilder: CaptureRequest.Builder
        private var outputSurfaces: MutableList<Surface> = ArrayList()
        private var autoFocus: Boolean = camera2Controller.getAutoFocus()
        private var flash: CameraFlash = camera2Controller.getFlash()
        private var autoWhiteBalance: Boolean = camera2Controller.getAutoWhiteBalance()
        private var cameraFacing: CameraFacing = camera2Controller.getFacing()
        private var displayOrientation = cameraPreview.displayOrientation
        fun setOutputSurface(outputSurface: Surface): CaptureRequestBuilder {
            outputSurfaces.add(outputSurface)
            return this
        }

        fun setOutputSurfaces(outputSurfaces: MutableList<Surface>): CaptureRequestBuilder {
            this.outputSurfaces = outputSurfaces
            return this
        }

        fun setAutoFocus(autoFocus: Boolean): CaptureRequestBuilder {
            this.autoFocus = autoFocus
            return this
        }

        fun setFlash(flash: CameraFlash): CaptureRequestBuilder {
            this.flash = flash
            return this
        }

        fun setAutoWhiteBalance(autoWhiteBalance: Boolean): CaptureRequestBuilder {
            this.autoWhiteBalance = autoWhiteBalance
            return this
        }

        fun setDisplayOrientation(
            cameraFacing: CameraFacing,
            displayOrientation: Int
        ): CaptureRequestBuilder {
            this.cameraFacing = cameraFacing
            this.displayOrientation = displayOrientation
            return this
        }

        fun <T> setKeyValue(key: CaptureRequest.Key<T>, value: T): CaptureRequestBuilder {
            previewRequestBuilder[key] = value
            return this
        }

        fun build(): CaptureRequest {
            setup3AControls(previewRequestBuilder, autoFocus, flash, autoWhiteBalance)
            setupJpegOrientation(previewRequestBuilder, cameraFacing, displayOrientation)
            for (surface in outputSurfaces) {
                previewRequestBuilder.addTarget(surface)
            }
            return previewRequestBuilder.build()
        }

        init {
            previewRequestBuilder =
                camera2Controller.cameraDevice.createCaptureRequest(CameraDevice.TEMPLATE_STILL_CAPTURE)
        }
    }
}